File: kotlin/src/main/kotlin/two_pointer/ContainerWithMostWater.kt
package org.example.two_pointer

import kotlin.math.max
import kotlin.math.min

fun maxArea(height: IntArray): Int {
    var left = 0
    var right = height.size - 1

    var res = 0
    while (left <= right) {
        val maxArea = min(height[left], height[right]) * (right - left)
        res = max(res, maxArea)

        if (height[left] < height[right]) {
            left++
        } else {
            right--
        }
    }

    return res
}

fun main() {
    val input = intArrayOf(1, 8, 6, 2, 5, 4, 8, 3, 7)
    val output = maxArea(input)
    println("output: $output, expected: 49")
}

File: python/leetcode/neetcode-roadmap/two_pointer/container_with_most_water.py
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        maxArea = 0
        left, right = 0, len(height) - 1
        while left < right:
            width = right - left
            area = width * min(height[left], height[right])
            maxArea = max(maxArea, area)
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
        return maxArea


s = Solution()
print(s.maxArea(height=[1, 8, 6, 2, 5, 4, 8, 3, 7]))
